File: skills/agents-pair/scripts/stream_monitor.py
#!/usr/bin/env python3
"""Mirror Claude stream-json to disk and print compact live progress."""

from __future__ import annotations

import json
import os
import sys
from typing import Any


def shorten(value: Any, limit: int = 240) -> str:
    if value is None:
        return ""
    if not isinstance(value, str):
        value = json.dumps(value, sort_keys=True)
    value = " ".join(value.replace("\r", " ").replace("\n", " ").split())
    if len(value) > limit:
        return value[: limit - 3] + "..."
    return value


def emit(message: str) -> None:
    print(message, file=sys.stderr, flush=True)


def describe_tool_input(name: str, tool_input: Any) -> str:
    if not isinstance(tool_input, dict):
        return shorten(tool_input)
    if name == "StructuredOutput":
        message = tool_input.get("message_to_codex")
        status = (tool_input.get("continuation_state") or {}).get("status")
        next_owner = (tool_input.get("proposed_next_turn") or {}).get("owner")
        return shorten(
            {
                "message_to_codex": message,
                "continuation_status": status,
                "next_owner": next_owner,
            }
        )
    if "command" in tool_input:
        return "command=" + shorten(tool_input.get("command"))
    if "file_path" in tool_input:
        return "file_path=" + shorten(tool_input.get("file_path"))
    if "path" in tool_input:
        return "path=" + shorten(tool_input.get("path"))
    if "pattern" in tool_input:
        return "pattern=" + shorten(tool_input.get("pattern"))
    return shorten(tool_input)


TEXT_LIMIT = int(os.environ.get("CLAUDE_MONITOR_TEXT_LIMIT", "3000"))
text_emitted = 0
text_limit_reported = False


def emit_text_delta(text: str) -> None:
    global text_emitted, text_limit_reported
    text = shorten(text, 400)
    if not text or TEXT_LIMIT <= 0:
        return
    remaining = TEXT_LIMIT - text_emitted
    if remaining <= 0:
        if not text_limit_reported:
            emit("[claude:text] live text limit reached; see stream transcript for full output")
            text_limit_reported = True
        return
    if len(text) > remaining:
        text = text[: max(0, remaining - 3)] + "..."
    text_emitted += len(text)
    emit(f"[claude:text] {text}")


def handle_event(obj: dict[str, Any]) -> None:
    event_type = obj.get("type")

    if event_type == "system":
        subtype = obj.get("subtype")
        if subtype == "init":
            tools = ",".join(obj.get("tools") or [])
            emit(
                "[claude:init] "
                f"session={obj.get('session_id')} model={obj.get('model')} tools={tools}"
            )
        elif subtype == "status":
            emit(f"[claude:status] {obj.get('status')}")
        elif subtype == "api_retry":
            emit(
                "[claude:retry] "
                f"attempt={obj.get('attempt')}/{obj.get('max_retries')} "
                f"status={obj.get('error_status')} error={obj.get('error')} "
                f"delay_ms={obj.get('retry_delay_ms')}"
            )
        elif subtype == "hook_response" and obj.get("outcome") != "success":
            emit(
                "[claude:hook] "
                f"{obj.get('hook_name')} outcome={obj.get('outcome')} "
                f"stderr={shorten(obj.get('stderr'))}"
            )
        return

    if event_type == "stream_event":
        event = obj.get("event") or {}
        nested_type = event.get("type")
        if nested_type == "content_block_start":
            block = event.get("content_block") or {}
            block_type = block.get("type")
            if block_type == "tool_use":
                emit(f"[claude:tool] start {block.get('name')}")
            elif block_type == "thinking":
                emit("[claude:thinking] hidden reasoning block started")
        elif nested_type == "content_block_delta":
            delta = event.get("delta") or {}
            if delta.get("type") == "text_delta":
                emit_text_delta(delta.get("text") or "")
        elif nested_type == "message_delta":
            usage = event.get("usage") or {}
            details = usage.get("output_tokens_details") or {}
            thinking_tokens = details.get("thinking_tokens")
            stop_reason = (event.get("delta") or {}).get("stop_reason")
            parts = []
            if thinking_tokens:
                parts.append(f"thinking_tokens={thinking_tokens}")
            if stop_reason:
                parts.append(f"stop_reason={stop_reason}")
            if parts:
                emit("[claude:progress] " + " ".join(parts))
        return

    if event_type == "assistant":
        message = obj.get("message") or {}
        for block in message.get("content") or []:
            if block.get("type") == "tool_use":
                name = block.get("name")
                emit(f"[claude:tool] {name} {describe_tool_input(name, block.get('input'))}")
            elif block.get("type") == "text" and obj.get("error"):
                emit(f"[claude:error] {shorten(block.get('text'), 500)}")
        return

    if event_type == "user":
        message = obj.get("message") or {}
        for block in message.get("content") or []:
            if block.get("type") == "tool_result":
                content = block.get("content")
                size = len(content) if isinstance(content, str) else 0
                emit(f"[claude:tool-result] {block.get('tool_use_id')} bytes={size}")
        return

    if event_type == "result":
        structured = obj.get("structured_output")
        if isinstance(structured, dict):
            status = (structured.get("continuation_state") or {}).get("status")
            next_owner = (structured.get("proposed_next_turn") or {}).get("owner")
            message = shorten(structured.get("message_to_codex"), 500)
            emit(
                "[claude:peer] "
                f"status={status} next_owner={next_owner} message={message}"
            )
        emit(
            "[claude:done] "
            f"terminal={obj.get('terminal_reason')} stop={obj.get('stop_reason')} "
            f"turns={obj.get('num_turns')} duration_ms={obj.get('duration_ms')}"
        )


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: stream_monitor.py RAW_STREAM_JSONL", file=sys.stderr)
        return 2

    with open(sys.argv[1], "w", encoding="utf-8") as raw:
        for line in sys.stdin:
            raw.write(line)
            raw.flush()
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue
            if isinstance(obj, dict):
                handle_event(obj)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/agents-pair/scripts/stream_monitor.mjs
#!/usr/bin/env node
// Mirror Claude stream-json to disk and print compact live progress.

import fs from "node:fs";

function pythonJsonString(value) {
  return JSON.stringify(value).replace(/[\u007f-\uffff]/g, (char) => {
    return `\\u${char.charCodeAt(0).toString(16).padStart(4, "0")}`;
  });
}

function stableStringify(value) {
  if (value === null || value === undefined) {
    return "null";
  }
  if (typeof value === "string") {
    return pythonJsonString(value);
  }
  if (typeof value === "number") {
    return Number.isFinite(value) ? String(value) : "null";
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(", ")}]`;
  }
  if (typeof value === "object") {
    const pairs = Object.keys(value)
      .sort()
      .map((key) => `${pythonJsonString(key)}: ${stableStringify(value[key])}`);
    return `{${pairs.join(", ")}}`;
  }
  return pythonJsonString(String(value));
}

function shorten(value, limit = 240) {
  if (value === null || value === undefined) {
    return "";
  }
  if (typeof value !== "string") {
    value = stableStringify(value);
  }
  value = value.replace(/\r/g, " ").replace(/\n/g, " ");
  value = value.trim() ? value.trim().split(/\s+/).join(" ") : "";
  if (value.length > limit) {
    return `${value.slice(0, limit - 3)}...`;
  }
  return value;
}

function emit(message) {
  process.stderr.write(`${message}\n`);
}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function get(obj, key) {
  return isObject(obj) && Object.hasOwn(obj, key) ? obj[key] : null;
}

function pyStr(value) {
  if (value === null || value === undefined) {
    return "None";
  }
  if (typeof value === "boolean") {
    return value ? "True" : "False";
  }
  return String(value);
}

function describeToolInput(name, toolInput) {
  if (!isObject(toolInput)) {
    return shorten(toolInput);
  }
  if (name === "StructuredOutput") {
    const continuationState = get(toolInput, "continuation_state") || {};
    const proposedNextTurn = get(toolInput, "proposed_next_turn") || {};
    const message = get(toolInput, "message_to_codex");
    const status = get(continuationState, "status");
    const nextOwner = get(proposedNextTurn, "owner");
    return shorten({
      message_to_codex: message,
      continuation_status: status,
      next_owner: nextOwner,
    });
  }
  if (Object.hasOwn(toolInput, "command")) {
    return `command=${shorten(get(toolInput, "command"))}`;
  }
  if (Object.hasOwn(toolInput, "file_path")) {
    return `file_path=${shorten(get(toolInput, "file_path"))}`;
  }
  if (Object.hasOwn(toolInput, "path")) {
    return `path=${shorten(get(toolInput, "path"))}`;
  }
  if (Object.hasOwn(toolInput, "pattern")) {
    return `pattern=${shorten(get(toolInput, "pattern"))}`;
  }
  return shorten(toolInput);
}

const textLimit = Number.parseInt(process.env.CLAUDE_MONITOR_TEXT_LIMIT || "3000", 10);
let textEmitted = 0;
let textLimitReported = false;

function emitTextDelta(text) {
  text = shorten(text, 400);
  if (!text || textLimit <= 0) {
    return;
  }
  const remaining = textLimit - textEmitted;
  if (remaining <= 0) {
    if (!textLimitReported) {
      emit("[claude:text] live text limit reached; see stream transcript for full output");
      textLimitReported = true;
    }
    return;
  }
  if (text.length > remaining) {
    text = `${text.slice(0, Math.max(0, remaining - 3))}...`;
  }
  textEmitted += text.length;
  emit(`[claude:text] ${text}`);
}

function handleEvent(obj) {
  const eventType = get(obj, "type");

  if (eventType === "system") {
    const subtype = get(obj, "subtype");
    if (subtype === "init") {
      const toolsValue = get(obj, "tools");
      const tools = Array.isArray(toolsValue) ? toolsValue.join(",") : "";
      emit(
        `[claude:init] session=${pyStr(get(obj, "session_id"))} ` +
          `model=${pyStr(get(obj, "model"))} tools=${tools}`,
      );
    } else if (subtype === "status") {
      emit(`[claude:status] ${pyStr(get(obj, "status"))}`);
    } else if (subtype === "api_retry") {
      emit(
        `[claude:retry] attempt=${pyStr(get(obj, "attempt"))}/${pyStr(get(obj, "max_retries"))} ` +
          `status=${pyStr(get(obj, "error_status"))} error=${pyStr(get(obj, "error"))} ` +
          `delay_ms=${pyStr(get(obj, "retry_delay_ms"))}`,
      );
    } else if (subtype === "hook_response" && get(obj, "outcome") !== "success") {
      emit(
        `[claude:hook] ${pyStr(get(obj, "hook_name"))} outcome=${pyStr(get(obj, "outcome"))} ` +
          `stderr=${shorten(get(obj, "stderr"))}`,
      );
    }
    return;
  }

  if (eventType === "stream_event") {
    const event = get(obj, "event") || {};
    const nestedType = get(event, "type");
    if (nestedType === "content_block_start") {
      const block = get(event, "content_block") || {};
      const blockType = get(block, "type");
      if (blockType === "tool_use") {
        emit(`[claude:tool] start ${pyStr(get(block, "name"))}`);
      } else if (blockType === "thinking") {
        emit("[claude:thinking] hidden reasoning block started");
      }
    } else if (nestedType === "content_block_delta") {
      const delta = get(event, "delta") || {};
      if (get(delta, "type") === "text_delta") {
        emitTextDelta(get(delta, "text") || "");
      }
    } else if (nestedType === "message_delta") {
      const usage = get(event, "usage") || {};
      const details = get(usage, "output_tokens_details") || {};
      const thinkingTokens = get(details, "thinking_tokens");
      const delta = get(event, "delta") || {};
      const stopReason = get(delta, "stop_reason");
      const parts = [];
      if (thinkingTokens) {
        parts.push(`thinking_tokens=${pyStr(thinkingTokens)}`);
      }
      if (stopReason) {
        parts.push(`stop_reason=${pyStr(stopReason)}`);
      }
      if (parts.length) {
        emit(`[claude:progress] ${parts.join(" ")}`);
      }
    }
    return;
  }

  if (eventType === "assistant") {
    const message = get(obj, "message") || {};
    const content = get(message, "content") || [];
    for (const block of content) {
      if (!isObject(block)) {
        continue;
      }
      if (get(block, "type") === "tool_use") {
        const name = get(block, "name");
        emit(`[claude:tool] ${pyStr(name)} ${describeToolInput(name, get(block, "input"))}`);
      } else if (get(block, "type") === "text" && get(obj, "error")) {
        emit(`[claude:error] ${shorten(get(block, "text"), 500)}`);
      }
    }
    return;
  }

  if (eventType === "user") {
    const message = get(obj, "message") || {};
    const content = get(message, "content") || [];
    for (const block of content) {
      if (!isObject(block)) {
        continue;
      }
      if (get(block, "type") === "tool_result") {
        const resultContent = get(block, "content");
        const size = typeof resultContent === "string" ? resultContent.length : 0;
        emit(`[claude:tool-result] ${pyStr(get(block, "tool_use_id"))} bytes=${size}`);
      }
    }
    return;
  }

  if (eventType === "result") {
    const structured = get(obj, "structured_output");
    if (isObject(structured)) {
      const continuationState = get(structured, "continuation_state") || {};
      const proposedNextTurn = get(structured, "proposed_next_turn") || {};
      const status = get(continuationState, "status");
      const nextOwner = get(proposedNextTurn, "owner");
      const message = shorten(get(structured, "message_to_codex"), 500);
      emit(`[claude:peer] status=${pyStr(status)} next_owner=${pyStr(nextOwner)} message=${message}`);
    }
    emit(
      `[claude:done] terminal=${pyStr(get(obj, "terminal_reason"))} ` +
        `stop=${pyStr(get(obj, "stop_reason"))} turns=${pyStr(get(obj, "num_turns"))} ` +
        `duration_ms=${pyStr(get(obj, "duration_ms"))}`,
    );
  }
}

function processLine(fd, line) {
  fs.writeSync(fd, line, undefined, "utf8");
  try {
    const obj = JSON.parse(line);
    if (isObject(obj)) {
      handleEvent(obj);
    }
  } catch {
    // Match the Python monitor: malformed stream lines are still mirrored.
  }
}

async function main() {
  if (process.argv.length !== 3) {
    process.stderr.write("usage: stream_monitor.mjs RAW_STREAM_JSONL\n");
    return 2;
  }

  const fd = fs.openSync(process.argv[2], "w");
  let buffer = "";
  process.stdin.setEncoding("utf8");
  try {
    for await (const chunk of process.stdin) {
      buffer += chunk;
      let newlineIndex = buffer.indexOf("\n");
      while (newlineIndex !== -1) {
        const line = buffer.slice(0, newlineIndex + 1);
        buffer = buffer.slice(newlineIndex + 1);
        processLine(fd, line);
        newlineIndex = buffer.indexOf("\n");
      }
    }
    if (buffer) {
      processLine(fd, buffer);
    }
  } finally {
    fs.closeSync(fd);
  }
  return 0;
}

process.exitCode = await main();
